### PR TITLE
Improve ses6 upgrade chapter (#843)

### LINKED ISSUE #843: Improve SES 6 upgrade documentation
**Link To Section**

https://documentation.suse.com/ses/6/single-html/ses-deployment/#cha-ceph-upgrade

**Summary**

A large number of customers will be upgrading from 5 -> 6 with the EOL happening in January. We need to clean up the docs to make this a more linear process with clear guidelines.

**Basic Example**

Nathan outlined basic steps to take:

1. structure the upgrade chapter as a series of steps
2. at the beginning of the chapter, explain that there are two methods of upgrading: DMS and ISO, what they are, what the difference between them is
3. label each step in the procedure so it's clear that the step applies to the "ISO method", "DMS method", or "both methods"

**Motivation**

It's not great.

**Notes**

See: https://github.com/SUSE/doc-ses/pull/842

### PR DESCRIPTION
**Pull request by tbazant**
Closes #843 
- only file xml/admin_ceph_upgrade.xml is affected, the rest is not relevant
- rendered version is attached 
[cha-ceph-upgrade_color_en.pdf](https://github.com/SUSE/doc-ses/files/5801027/cha-ceph-upgrade_color_en.pdf)

### CODE REVIEW

File: xml/admin_ceph_upgrade.xml [hunk 1]
```diff
@@ -203,340 +59,448 @@
...
+     </para>
+<screen>/srv/pillar/ceph/proposals/config/stack/default/ceph/cluster.yml</screen>
+     <para>
+      If <option>configuration_init</option> is still set to
+      <option>default-import</option>, the cluster is using
+      <filename>ceph.conf.import</filename> as its configuration file and not
+      &deepsea;'s default <filename>ceph.conf</filename> which is compiled from
+      files in
+     </para>
+<screen>/srv/salt/ceph/configuration/files/ceph.conf.d/</screen>
```
Comment (lproven) [inline xml/admin_ceph_upgrade.xml:138]: Ditto

File: xml/admin_ceph_upgrade.xml [hunk 2]
```diff
@@ -203,340 +59,448 @@
...
+     <para>
+      Therefore you need to inspect <filename>ceph.conf.import</filename> for
+      any custom configuration, and possibly move the configuration to one of
+      the files in
+     </para>
+<screen>/srv/salt/ceph/configuration/files/ceph.conf.d/</screen>
```
Comment (lproven) [inline xml/admin_ceph_upgrade.xml:144]: Ditto

File: xml/admin_ceph_upgrade.xml [hunk 3]
```diff
@@ -203,340 +59,448 @@
...
+     <para>
+      Then remove the <option>configuration_init: default-import</option> line
+      from
+     </para>
+<screen>/srv/pillar/ceph/proposals/config/stack/default/ceph/cluster.yml</screen>
```
Comment (lproven) [inline xml/admin_ceph_upgrade.xml:149]: Ditto

File: xml/admin_ceph_upgrade.xml [hunk 4]
```diff
@@ -203,340 +59,448 @@
...
+     <warning>
+      <title>Default &deepsea; Configuration</title>
       <para>
-       &osd;s
+       If you <emphasis role="bold">do not</emphasis> merge the configuration
+       from <filename>ceph.conf.import</filename> and remove the
```
Comment (lproven) [inline xml/admin_ceph_upgrade.xml:154]: **If** I understand this correctly, would it be clearer to also say "and do not remove the blah option"?

File: xml/admin_ceph_upgrade.xml [hunk 5]
```diff
@@ -203,340 +59,448 @@
...
+       will not be applied to the cluster.
       </para>
-     </listitem>
-     <listitem>
+     </warning>
+    </listitem>
+    <listitem>
+     <para>
+      Run the <command>salt-run upgrade.check</command> command to verify that
+      the cluster uses the new bucket type <literal>straw2</literal> and that
```
Comment (lproven) [inline xml/admin_ceph_upgrade.xml:165]:
```suggestion
      the cluster uses the new bucket type <literal>straw2</literal>, and that
```

File: xml/admin_ceph_upgrade.xml [hunk 6]
```diff
@@ -203,340 +59,448 @@
...
+   <title>Adjust &aa;</title>
+   <para>
+    If you used &aa; in either 'complain' or 'enforce' mode, you need to set a
+    &spillar; variable before upgrading. Because &cephos; ships with &aa; by
+    default, &aa; management was integrated into &deepsea; stage 0. The default
+    behavior in &productname; &productnumber; is to remove &aa; and related
+    profiles. If you want to retain the behavior configured in &productname;
+    &prevproductnumber;, verify that one of the following lines is present in
+    the <filename>/srv/pillar/ceph/stack/global.yml</filename> file before
+    starting the upgrade:
```
Comment (lproven) [inline xml/admin_ceph_upgrade.xml:228]: Do we want the product-number to be a self-updating thing? It seems to me that a literal product number would be preferable here, but I could be wrong.

File: xml/admin_ceph_upgrade.xml [hunk 7]
```diff
@@ -203,340 +59,448 @@
...
-     status</command> command, or by restarting an MDS and checking its logs
-     for the following message:
-    </para>
+  </sect2>
+
+  <sect2 xml:id="before-upgrade-mds-names">
+   <title>Verify MDS Names</title>
+   <para>
+    From &productname; &productnumber;, MDS names starting with a digit are no
+    longer allowed and MDS daemons will refuse to start. You can check whether
```
Comment (lproven) [inline xml/admin_ceph_upgrade.xml:245]:
```suggestion
    From &productname; &productnumber;, MDS names are no longer allowed to begin with a
    digit, and such names will cause MDS daemons to refuse to start. You can check whether
```

File: xml/admin_ceph_upgrade.xml [hunk 8]
```diff
@@ -203,340 +59,448 @@
...
 a future version.  MDS names may not start with a numeric digit.
 </screen>
-    <para>
-     If you see the above message, the MDS names will need to be migrated
-     before attempting to upgrade to &productname; &productnumber;. &deepsea;
-     provides an orchestration to automate such a migration. MDS names starting
-     with a digit will be prepended with 'mds.':
-    </para>
+   <para>
+    If you see the above message, the MDS names will need to be migrated before
```
Comment (lproven) [inline xml/admin_ceph_upgrade.xml:255]:
```suggestion
    If you see the above message, the MDS names must be migrated before
```

File: xml/admin_ceph_upgrade.xml [hunk 9]
```diff
@@ -203,340 +59,448 @@
...
-   <systemitem>chronyd</systemitem> either
-   <emphasis role="bold">before</emphasis> the cluster, or upgrade the cluster
-   and migrate to <systemitem class="daemon">chronyd</systemitem>
-   <emphasis role="bold">afterward</emphasis>.
-  </para>
+   </tip>
+   <para>
+    This will add MDS daemons with the new names before removing the old MDS
+    daemons. The number of MDS daemons will double for a short time. Clients
+    will be able to access &cephfs; only after a short pause to failover.
```
Comment (lproven) [inline xml/admin_ceph_upgrade.xml:291]:
```suggestion
    will be able to access &cephfs; only after a short pause for failover to happen.
```

Comment (lproven) [reply xml/admin_ceph_upgrade.xml:291]: This is only a suggestion -- I am not 100% what was meant.

File: xml/admin_ceph_upgrade.xml [hunk 10]
```diff
@@ -203,340 +59,448 @@
...
+    Therefore plan the migration for times when you expect little or no
```
Comment (lproven) [inline xml/admin_ceph_upgrade.xml:292]: Surely "for a time" would make more sense?

File: xml/admin_ceph_upgrade.xml [hunk 11]
```diff
@@ -545,309 +509,445 @@ mds standby for name = mds.456-another-mds
...
+       &prevcephos; to each &ceph; cluster node. Verify that correct software
+       repositories are connected to each cluster node (see
+       <xref linkend="upgrade-prepare-repos" />) and run &deepsea; stage 0:
+      </para>
+<screen>&prompt.smaster;salt-run state.orch ceph.stage.0</screen>
+     </step>
+     <step>
+      <para>
+       After stage 0 completes, verify that each cluster node's status includes
+       'HEALTH_OK'. If not, resolve the problem before possible reboots in next
```
Comment (lproven) [inline xml/admin_ceph_upgrade.xml:611]:
```suggestion
       'HEALTH_OK'. If not, resolve the problem before any possible reboots in
       the next
```

File: xml/admin_ceph_upgrade.xml [hunk 12]
```diff
@@ -545,309 +509,445 @@ mds standby for name = mds.456-another-mds
...
-      SUSE-Enterprise-Storage-5-Updates
+      Use <command>supportconfig</command> to collect important system
+      information and save it outside cluster nodes. Find more information in
+      <link xlink:href="https://documentation.suse.com/sles/12-SP5/single-html/SLES-admin/#sec-admsupport-supportconfig"/>.
+     </para>
+    </listitem>
+    <listitem>
+     <para>
+      Ensure there is enough free disk space on each cluster node. Check free
+      disk space with <command>df -h</command>. When needed, free disk space by
```
Comment (lproven) [inline xml/admin_ceph_upgrade.xml:690]:
```suggestion
      disk space with <command>df -h</command>. When needed, free up additional disk space by
```

File: xml/admin_ceph_upgrade.xml [hunk 13]
```diff
@@ -545,309 +509,445 @@ mds standby for name = mds.456-another-mds
...
+      removing unneeded files/directories or removing obsolete OS snapshots. If
+      there is not enough free disk space, do not continue with the upgrade
+      until having freed enough disk space.
```
Comment (lproven) [inline xml/admin_ceph_upgrade.xml:693]:
```suggestion
      until you have freed enough disk space.
```

File: xml/admin_ceph_upgrade.xml [hunk 14]
```diff
@@ -545,309 +509,445 @@ mds standby for name = mds.456-another-mds
...
     <para>
-     Document specific parts of the system setup, such as network setup,
-     partitioning, or installation details.
+     Rebuild the nodes before continuing:
     </para>
-   </listitem>
-   <listitem>
+<screen>
+ &prompt.smaster; salt-run rebuild.node <replaceable>TARGET</replaceable>
+ </screen>
```
Comment (lproven) [inline xml/admin_ceph_upgrade.xml:869]:
```suggestion
<screen>&prompt.smaster; salt-run rebuild.node <replaceable>TARGET</replaceable>/screen>
```

File: xml/admin_ceph_upgrade.xml [hunk 15]
```diff
@@ -904,7 +1004,7 @@ mds standby for name = mds.456-another-mds
 </screen>
  </sect1>
  <sect1 xml:id="upgrade-one-node">
-  <title>Per Node Upgrade&mdash;Basic Procedure</title>
+  <title>Per Node Upgrade Instructions</title>
```
Comment (lproven) [inline xml/admin_ceph_upgrade.xml:1007]:
```suggestion
  <title>Per-Node Upgrade Instructions</title>
```

Comment (lproven) [reply xml/admin_ceph_upgrade.xml:1007]: Also see my earlier comment about sentence case.

**Review by lproven — CHANGES_REQUESTED** [review]

File: xml/tuning-os-tuning.xml [hunk 16]
```diff
@@ -14,56 +14,61 @@
...
+  for significant performance improvement at the time of this work.
  </para>
  <sect1 xml:id="tuning-sles">
-   <title>&sle; Install and Validation of Base Performance</title>
+  <title>&sle; Install and Validation of Base Performance</title>
+
+  <para>
+   During the OS installation, do <emphasis>not</emphasis> select an install
+   pattern that includes an X-server. Doing so utilizes RAM and CPU resources
+   that are better allocated to tuning storage related daemons. We recommend a
```
Comment (lproven) [inline xml/tuning-os-tuning.xml:31]:
```suggestion
   that would be better allocated to tuning storage-related daemons. We recommend a
```

File: xml/tuning-os-tuning.xml [hunk 17]
```diff
@@ -14,56 +14,61 @@
...
+
+  <sect2>
+   <title>Network Performance</title>
+   <para>
+    To perform <command>iperf3</command> tests for network performance,
+    consider increasing the window size (<option>-w</option>) and running
+    multiple streams to fully test the bandwidth capability. If at standard
+    MTU, the NIC should be capable of running at approximately 70-80% of the
+    advertised bandwidth. Move up to jumbo frames, and the NIC should be able
+    to saturate the line.
```
Comment (lproven) [inline xml/tuning-os-tuning.xml:57]:
```suggestion
    To perform <command>iperf3</command> tests for network performance,
    consider increasing the window size (with the <option>-w</option> option)
    and running multiple streams to fully test the bandwidth capability. If using
    the standard MTU, your NICs should be capable of running at approximately
    70-80% of the advertised bandwidth. If you move up to jumbo frames, the
    NIC should be able to saturate the link.
```

File: xml/tuning-os-tuning.xml [hunk 18]
```diff
@@ -14,56 +14,61 @@
...
+   </para>
+   <para>
+    This is not necessarily true for faster topologies like 100&nbsp;Gb. In
+    those topologies, saturating the NIC can take a substantial amount of OS
+    and driver tuning in combination with ensuring the hardware has the
+    appropriate CPU clock speeds and settings.
```
Comment (lproven) [inline xml/tuning-os-tuning.xml:63]:
```suggestion
    This is not necessarily true for faster topologies such as 100&nbsp;Gb. In
    those topologies, saturating the NIC can require substantial OS and driver
    tuning, in combination with ensuring the hardware has the appropriate
    CPU clock speeds and settings.
```

File: xml/tuning-os-tuning.xml [hunk 19]
```diff
@@ -14,56 +14,61 @@
...
+   </para>
+   <para>
+    This is a sample of the <command>iperf3</command> commands used on a
+    100&nbsp;Gb network. In the command line, the <option>-N</option> disables
+    Nagle's buffering algorithm and the <option>-l</option> sets the buffer
+    length to higher than the default 128k, resulting in slightly higher
```
Comment (lproven) [inline xml/tuning-os-tuning.xml:69]:
```suggestion
    length to higher than the default 128&nbsp;kB, resulting in slightly higher
```

File: xml/tuning-os-tuning.xml [hunk 20]
```diff
@@ -86,244 +91,254 @@
...
+
+  <sect2>
+   <title>Storage Performance</title>
+   <para>
+    Use <command>fio</command> to test individual storage devices to understand
+    the per-device performance maxima. Do this for all devices to ensure there
+    are not any that are out of spec or are connected to expanders that lower
+    bandwidth. Doing an exhaustive study of different I/O sizes and patterns
+    would provide the most information about performance expectations, but is
+    beyond the scope of this document.
```
Comment (lproven) [inline xml/tuning-os-tuning.xml:104]:
```suggestion
    Use <command>fio</command> to test individual storage devices to understand
    the per-device performance maxima. Do this for all devices to ensure that none
    are out of specification, or are connected to expanders which lower
    bandwidth. Doing an exhaustive study of different I/O sizes and patterns
    would provide the most information about performance expectations, but is
    beyond the scope of this document.
```

File: xml/tuning-os-tuning.xml [hunk 21]
```diff
@@ -86,244 +91,254 @@
...
+   </para>
+   <para>
+    We recommend testing at least random 4&nbsp;kB, random 64&nbsp;kB, and
+    sequential 64&nbsp;kB and 1&nbsp;MB. This gives a reasonable overall view
```
Comment (lproven) [inline xml/tuning-os-tuning.xml:108]:
```suggestion
    sequential 64&nbsp;kB and 1&nbsp;MB buffers. This should give a reasonable overall view
```

File: xml/tuning-os-tuning.xml [hunk 22]
```diff
@@ -86,244 +91,254 @@
...
+    important to use the raw device
+    (<literal>/dev/sd<replaceable>X</replaceable></literal>) and to use the
+    <option>direct=1</option> option with multiple jobs to maximize device
+    performance under stress.
+   </para>
+   <para>
+    Make sure that the test size (dataset) is large enough to over-run any
+    caches that may apply. We recommend using the <literal>ramp-time</literal>
+    parameter to allow for enough time for the cache overrun to occur before
+    performance measurements are made, helping to ensure that performance
```
Comment (lproven) [inline xml/tuning-os-tuning.xml:119]:
```suggestion
    performance measurements are made. This helps to ensure that performance
```

File: xml/tuning-os-tuning.xml [hunk 23]
```diff
@@ -86,244 +91,254 @@
...
+    numbers are not tainted by cache-only performance.
+   </para>
+   <para>
+    Run <command>fio</command> against all devices in an OSD node
+    simultaneously to identify bottlenecks. It should scale very near linearly;
+    if not, check controller firmware, slot placement, and if necessary, split
+    devices across multiple controllers. This is simulating the node under
```
Comment (lproven) [inline xml/tuning-os-tuning.xml:126]:
```suggestion
    devices across multiple controllers. This simulates the node under
```

File: xml/tuning-os-tuning.xml [hunk 24]
```diff
@@ -86,244 +91,254 @@
...
+    heavy load.
+   </para>
+   <para>
+    We recommend to use the same I/O patterns and block sizes that the
```
Comment (lproven) [inline xml/tuning-os-tuning.xml:130]:
```suggestion
    We recommend using the same I/O patterns and block sizes that the
```

File: xml/tuning-os-tuning.xml [hunk 25]
```diff
@@ -86,244 +91,254 @@
...
+    individual devices were tested with. The job count should be a multiple of
+    the total number of devices in the system to allow for even distribution
```
Comment (lproven) [inline xml/tuning-os-tuning.xml:132]:
```suggestion
    the total number of devices in the system, to allow for even distribution
```

File: xml/tuning-os-tuning.xml [hunk 26]
```diff
@@ -86,244 +91,254 @@
...
+    across all devices and buses.
+   </para>
+   <sect3>
+    <title>Latency Bound and Maximum</title>
+    <para>
+     There is value in performing both latency-bound and worst-case-scenario
+     tests. Changing a particular value may not improve latency, but rather
+     enable the cluster to handle even more total load, even though latencies
+     continue to rise. The inverse may also be true where a change affects
```
Comment (lproven) [inline xml/tuning-os-tuning.xml:141]:
```suggestion
     continue to rise. The inverse may also be true, where a change affects
```

File: xml/tuning-os-tuning.xml [hunk 27]
```diff
@@ -86,244 +91,254 @@
...
-         maintain a 10&nbsp;ms average. It then backs down the queue depth
-         until the latency average is maintained.
-       </para>
-     </sect3>
-   </sect2>
+    <para>
+     The settings above cause <command>fio</command> to increase the I/O queue
+     depth until 1% or more of IOPS during a sliding five-second window no
+     longer maintain a 10&nbsp;ms average. It then backs down the queue depth
+     until the latency average is maintained.
```
Comment (lproven) [inline xml/tuning-os-tuning.xml:155]:
```suggestion
     The settings above cause <command>fio</command> to increase the I/O queue
     depth until 1% or more of IOPS no longer maintain a 10&nbsp;ms average during
     a sliding five-second window. It then backs down the queue depth until the
     latency average is maintained.
```

File: xml/tuning-os-tuning.xml [hunk 28]
```diff
@@ -86,244 +91,254 @@
...
+  <para>
+   There are several aspects of the kernel that can be tuned on both the
+   cluster and some of the clients. It is important to understand that most
+   tuning is about gaining very small incremental improvements that in
+   aggregate represent a measureable, and hopefully, meaningful improvement in
+   performance. Information on tuning comes from a variety of sources for this
+   work. The primary source is
+   <link xlink:href="https://documentation.suse.com/sles/15-SP1/html/SLES-all/book-sle-tuning.html"/>.
+   Numerous other references were utilized, including documentation from
+   hardware vendors.
```
Comment (lproven) [inline xml/tuning-os-tuning.xml:172]:
```suggestion
   There are several aspects of the kernel that can be tuned on both the
   cluster and some of the clients. It is important to understand that most
   tuning is about gaining very small incremental improvements, which in
   aggregate represent a measureable (and hopefully meaningful) improvement in
   performance. For this document, information on tuning comes from a variety of
   sources. The primary source is
   <link xlink:href="https://documentation.suse.com/sles/15-SP1/html/SLES-all/book-sle-tuning.html"/>.
   We also used numerous other references, including documentation from
   hardware vendors.
```

File: xml/tuning-os-tuning.xml [hunk 29]
```diff
@@ -86,244 +91,254 @@
...
-     There are several aspects of the kernel that can be tuned on both
-     the cluster and some of the clients. It is important to understand
-     that most tuning is about gaining very small incremental improvements
-     that in aggregate represent a measureable, and hopefully, meaningful
-     improvement in performance.
-     Information on tuning comes from a variety of sources for this work.
-     The primary source is <link xlink:href="https://documentation.suse.com/sles/15-SP1/html/SLES-all/book-sle-tuning.html"/>.
-     Numerous other references were utilized, including documentation
-     from hardware vendors.
+    One key area of kernel for performance tuning is to disable the
```
Comment (lproven) [inline xml/tuning-os-tuning.xml:178]:
```suggestion
    One key area of kernel performance tuning is to disable its
```

File: xml/tuning-os-tuning.xml [hunk 30]
```diff
@@ -86,244 +91,254 @@
...
+    side-channel attack mitigations present in the kernel. The bulk of the
```
Comment (lproven) [inline xml/tuning-os-tuning.xml:179]:
```suggestion
    side-channel attack mitigations. The bulk of the
```

File: xml/tuning-os-tuning.xml [hunk 31]
```diff
@@ -86,244 +91,254 @@
...
+    benefits from this tuning occur with smaller I/Os ranging in size from
```
Comment (lproven) [inline xml/tuning-os-tuning.xml:180]:
```suggestion
    benefits from this occur with smaller I/Os, ranging in size from
```

File: xml/tuning-os-tuning.xml [hunk 32]
```diff
@@ -86,244 +91,254 @@
...
+    4&nbsp;kB to 64&nbsp;kB. In particular, 64&nbsp;kB random read and
+    sequential writes doubled in performance in a limited test environment
+    using only two client nodes.
+   </para>
+   <para>
+    Changing these options requires a clear understanding of the security
+    implications as they involve disabling mitigations for side-channel attacks
```
Comment (lproven) [inline xml/tuning-os-tuning.xml:187]:
```suggestion
    implications, as they involve disabling mitigations for side-channel attacks
```

File: xml/tuning-os-tuning.xml [hunk 33]
```diff
@@ -86,244 +91,254 @@
...
+    on some CPUs. The need to disable these mitigations may be minimized with
+    newer processors. Evaluate carefully whether this is something needed for
```
Comment (lproven) [inline xml/tuning-os-tuning.xml:189]:
```suggestion
    newer processors. You must carefully evaluate whether this is something needed for
```

File: xml/tuning-os-tuning.xml [hunk 34]
```diff
@@ -86,244 +91,254 @@
...
+    the particular hardware being utilized. In the test configuration, a salt
```
Comment (lproven) [inline xml/tuning-os-tuning.xml:190]:
```suggestion
    the particular hardware being utilized. In the test configuration, a &salt;
```

File: xml/tuning-os-tuning.xml [hunk 35]
```diff
@@ -86,244 +91,254 @@
...
+    each customer environment. An example of adjusting the
+    <literal>grub2</literal> configuration file can also be found in the
+    appendix.
+   </para>
+  </sect2>
+
+  <sect2>
+   <title>I/O tuning - Multiqueue Block I/O</title>
+   <para>
+    The first aspect to tune is to ensure that I/O if flowing in the most
```
Comment (lproven) [inline xml/tuning-os-tuning.xml:211]:
```suggestion
    The first aspect to tune is to ensure that I/O is flowing in the most
```

File: xml/tuning-os-tuning.xml [hunk 36]
```diff
@@ -86,244 +91,254 @@
...
+    optimal pattern. For the test cluster used in this test, that means enable
```
Comment (lproven) [inline xml/tuning-os-tuning.xml:212]:
```suggestion
    optimal pattern. For the test cluster used in this test, that means enabling
```

File: xml/tuning-os-tuning.xml [hunk 37]
```diff
@@ -86,244 +91,254 @@
...
+    multi-queue block I/O. This is done through adding a boot-time kernel
+    parameter as found in the Section 12.4
```
Comment (lproven) [inline xml/tuning-os-tuning.xml:214]:
```suggestion
    parameter, as found in Section 12.4
```

File: xml/tuning-os-tuning.xml [hunk 38]
```diff
@@ -86,244 +91,254 @@
...
+    <link xlink:href="https://documentation.suse.com/sles/15-SP1/html/SLES-all/cha-tuning-io.html#cha-tuning-io-barrier"/>.
+    This is not an action that should be taken unilaterally on clusters that
+    contain spinning media devices due to potential performance degradation for
+    those devices. The general result is that there are multiple I/O queues
+    assigned to the device, allowing more jobs to be handled simultaneously by
+    those such as NVMe and SSD that can service large numbers of requests.
```
Comment (lproven) [inline xml/tuning-os-tuning.xml:220]:
```suggestion
    This is not an action that should be taken unilaterally on clusters which
    contain spinning media devices, due to potential performance degradation for
    those devices. The general result is that there multiple I/O queues are
    assigned to each device, allowing more jobs to be handled simultaneously by
    those device that can service large numbers of requests, such as  SSD and 
    NVMe drives.
```

File: xml/tuning-os-tuning.xml [hunk 39]
```diff
@@ -86,244 +91,254 @@
...
+   <para>
+    To get the best read performance, it may be necessary to adjust the
+    <option>read_ahead</option> and write cache settings for the SSD devices.
+    In our particular testing environment, disabling write cache and forcing
+    <option>read_ahead</option> to 2&nbsp;MB resulted in the best overall
+    performance.
+   </para>
+   <para>
+    Before tuning, it is important to check the default values and measure the
+    differences in performance compared to the baseline.
```
Comment (lproven) [inline xml/tuning-os-tuning.xml:235]:
```suggestion
    Before tuning, it is important to check and record the default values, and
    measure any differences in performance compared to that baseline.
```

File: xml/tuning-os-tuning.xml [hunk 40]
```diff
@@ -86,244 +91,254 @@
...
+   </para>
+   <para>
+    By placing the following file in <filename>/etc/udev/rules.d</filename> the
+    device is detected by the model name shown in
+    <filename>/sys/block/{devname}/device/model</filename> and assigned to
+    disable write caching and setting the <option>read_ahead_kb</option> to
+    2&nbsp;MB.
```
Comment (lproven) [inline xml/tuning-os-tuning.xml:242]:
```suggestion
    By placing the following file in <filename>/etc/udev/rules.d</filename>,
    devices will be detected according to the model name shown in
    <filename>/sys/block/{devname}/device/model</filename>, and instructed to
    disable write caching and set the <option>read_ahead_kb</option> option to
    2&nbsp;MB.
```

Comment (lproven) [reply xml/tuning-os-tuning.xml:242]: I note that no filename seems to be specified here. I think we should at least offer an example filename.

File: xml/tuning-os-tuning.xml [hunk 41]
```diff
@@ -86,244 +91,254 @@
...
-       Proper tuning of the network stack can substantially add to improving the
-       latency and throughput of the cluster. A full script for the testing
-       we performed can be found in <xref linkend="tuning-appendix-c"/>.
-     </para>
+  </sect2>
+
+  <sect2>
+   <title>Network Stack and Device Tuning</title>
+   <para>
+    Proper tuning of the network stack can substantially add to improving the
```
Comment (lproven) [inline xml/tuning-os-tuning.xml:256]:
```suggestion
    Proper tuning of the network stack can substantially assist improving the
```

File: xml/tuning-os-tuning.xml [hunk 42]
```diff
@@ -86,244 +91,254 @@
...
+    latency and throughput of the cluster. A full script for the testing we
+    performed can be found in <xref linkend="tuning-appendix-c"/>.
+   </para>
+   <para>
+    The first and most impactful tuning is to utilize jumbo frame packets. For
```
Comment (lproven) [inline xml/tuning-os-tuning.xml:261]:
```suggestion
    The first change, and one with the highest impact, is to utilize jumbo frame packets. For
```

File: xml/tuning-os-tuning.xml [hunk 43]
```diff
@@ -86,244 +91,254 @@
...
+    this to be done, all interfaces utilizing the cluster must be set to the
```
Comment (lproven) [inline xml/tuning-os-tuning.xml:262]: This sounds strange to me.

Would "all interfaces used in the cluster" be better?

If it does not refer to internal comms between cluster nodes, then:

"all interfaces used to access the cluster"... but this seems less likely to me.

File: xml/tuning-os-tuning.xml [hunk 44]
```diff
@@ -86,244 +91,254 @@
...
+    same setting with an MTU of 9000. Network switches often have the ports set
+    to 9100 or higher. This is suitable as they are only passing packets, not
+    creating them.
```
Comment (lproven) [inline xml/tuning-os-tuning.xml:265]:
```suggestion
    this to be done, all interfaces utilizing the cluster must be set to use an MTU of
    9000 bytes. Network switches are often set to use 9100 or higher. This is 
    suitable, as they are only passing packets, not creating them.
```

File: xml/tuning-os-tuning.xml [hunk 45]
```diff
@@ -86,244 +91,254 @@
...
+      To set this persistently, utilize &yast; to set the MTU for the bonded
+      interface.
+     </para>
+    </sect4>
+    <sect4>
+     <title>PCIe Bus Adjustment</title>
+     <para>
+      Adjusting the PCIe maximum read request size can provide a slight boost
+      to performance. Be aware that this tuning is card- and slot-specific and
+      needs to only be done only in conjunction with the conditions and
```
Comment (lproven) [inline xml/tuning-os-tuning.xml:289]:
```suggestion
      must only be done in conjunction with the conditions and
```

Comment (lproven) [reply xml/tuning-os-tuning.xml:289]: Too many "only"s!

File: xml/tuning-os-tuning.xml [hunk 46]
```diff
@@ -86,244 +91,254 @@
...
-           <para>
-            The current and maximum settings can be revealed by issuing
-            the following command to the proper NICs:
-          </para>
+    </sect4>
+    <sect4>
+     <title>Ring Buffers</title>
+     <para>
+      Many NIC drivers start with a default value for the receive (RX) and
+      transmit (TX) buffers that is not optimal for high-throughput scenarios
```
Comment (lproven) [inline xml/tuning-os-tuning.xml:328]:
```suggestion
      transmit (TX) buffers that is not optimal for high-throughput scenarios,
```

File: xml/tuning-os-tuning.xml [hunk 47]
```diff
@@ -337,124 +352,126 @@ ethtool -g eth4
...
 </screen>
-          <para>
-            Here we can see that the NIC can allocate up to 8&nbsp;kB, but is
-            only currently using 1&nbsp;kB buffers. To adjust this for the
-            cluster, issue the following command:
-          </para>
+     <para>
+      Here we can see that the NIC can allocate up to 8&nbsp;kB, but is only
+      currently using 1&nbsp;kB buffers. To adjust this for the cluster, issue
+      the following command:
```
Comment (lproven) [inline xml/tuning-os-tuning.xml:358]:
```suggestion
      Here we can see that the NIC can allocate buffers of up to 8&nbsp;kB,
      but it iscurrently using ones of only 1&nbsp;kB. To adjust this for the cluster, issue
      the following command:
```

File: xml/tuning-os-tuning.xml [hunk 48]
```diff
@@ -337,124 +352,126 @@ ethtool -g eth4
...
+      value.
+     </para>
+    </sect4>
+   </sect3>
+   <sect3>
+    <title>Network Stack</title>
+    <para>
+     The following settings can all be made persistent by modifying
+     <filename>/etc/sysctl.conf</filename>. They are represented as arguments
+     in a salt command to allow testing and validation in your environment
```
Comment (lproven) [inline xml/tuning-os-tuning.xml:390]:
```suggestion
     in a &salt; command to allow testing and validation in your environment
```

File: xml/tuning-os-tuning.xml [hunk 49]
```diff
@@ -337,124 +352,126 @@ ethtool -g eth4
...
+     before making them permanent.
+    </para>
+    <sect4>
+     <title>Lower TCP Latency</title>
+     <para>
+      Setting the <literal>TCP low latency</literal> option disables IPv4 TCP
+      pre-queue processing and improves latency. We recommend experimenting
+      with this setting at both <literal>0</literal> and <literal>1</literal>.
```
Comment (lproven) [inline xml/tuning-os-tuning.xml:398]:
```suggestion
      with setting this to both <literal>0</literal> and <literal>1</literal>.
```

File: xml/tuning-os-tuning.xml [hunk 50]
```diff
@@ -337,124 +352,126 @@ ethtool -g eth4
...
-            were added to help ensure that packet sequence could be better tracked:
-          </para>
+    </sect4>
+    <sect4>
+     <title>TCP Sequence and Timestamps</title>
+     <para>
+      In fast networks, TCP sequence numbers can be re-used in a very short
+      timeframe. The result is that the system thinks a packet has been
+      received out of order, resulting in a drop. TCP timestamps were added to
+      help ensure that packet sequence could be better tracked:
```
Comment (lproven) [inline xml/tuning-os-tuning.xml:431]:
```suggestion
      help ensure that packet sequence could be tracked better:
```

**Review by lproven — CHANGES_REQUESTED** [review]

**Review by tbazant — COMMENTED** [review]
suggestions accepted

File: xml/tuning-os-tuning.xml (hunk 43 above)
Comment (tbazant) [inline xml/tuning-os-tuning.xml]:
```suggestion
    this to be done, all interfaces used in the cluster must be set to the
```

**Review by smithfarm — APPROVED** [review]
While reviewing this I found an additional problem, but it's not serious and I don't want to delay merge of this PR, so I'll open a new issue for it.

Looks good to me, and thanks @tbazant for doing the heavy lifting here.

**Comment by smithfarm** [comment]
Do we need to be concerned about the failed Travis CI builds?

**Review by lproven — APPROVED** [review]

**Pull request merged**